File: Range Sum Query 2D - Immutable.cpp
class NumMatrix {
public:
    vector<int> sums;
    vector<vector<int>> nums;
    NumMatrix(vector<vector<int>>& matrix) {
        nums = matrix;
        int a = nums.size(); 
        int b = nums[0].size();
        for(int i = 0; i<a; i++){
           sums.push_back(nums[i][0]);
            for(int j=1; j<b; j++){
                sums.push_back(sums.back() + nums[i][j]);
            }
        }
    }
    int sumRegion(int row1, int col1, int row2, int col2) {
        int a = nums.size();
        int b = nums[0].size();
        int answer = 0;
        int r = row1;
        while(r<=row2){
            answer += sums[(r*b)+col2] - ((col1>0)? sums[(r*b)+(col1-1)] : 0);
            r++;
        }
        return answer;
    }
};

File: Range Sum Query 2D - Immutable.py
class NumMatrix:
    matrix = [[]]
    def __init__(self, matrix: List[List[int]]):
        for col in range(1, len(matrix[0])):
            matrix[0][col] += matrix[0][col-1]
        for row in range(1, len(matrix)):
            matrix[row][0] += matrix[row-1][0]
        for row in range(1, len(matrix)):
            for col in range(1, len(matrix[0])):
                matrix[row][col] += (matrix[row-1][col] + matrix[row][col-1] - matrix[row-1][col-1])

        self.matrix = matrix

    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        sum = self.matrix[row2][col2]
        if row1 != 0:
            sum -= self.matrix[row1-1][col2]
            if col1 != 0:
                sum += self.matrix[row1-1][col1-1]
        if col1 != 0:
            sum -= self.matrix[row2][col1-1]
        
        return sum


# Your NumMatrix object will be instantiated and called as such:
# obj = NumMatrix(matrix)
# param_1 = obj.sumRegion(row1,col1,row2,col2)
